File: 백준/Silver/1343. 폴리오미노/폴리오미노.py
import sys

board = sys.stdin.readline()
ans = ""
cnt = 0

for i in range(len(board)):
    if board[i] == 'X':
        cnt += 1
    else:
        if (cnt % 2 == 0):
            if (cnt >= 4):
                for j in range(cnt // 4):
                    ans += "AAAA"
                    cnt -= 4
            for j in range(cnt // 2):
                ans += "BB"
                cnt -= 2
            if (i != len(board) - 1):
                ans += "."
        else:
            print(-1)
            exit()
print(ans)

File: 백준/Silver/1343. 폴리오미노/폴리오미노.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String board = br.readLine();
        board += " ";
        ArrayList<String> arr = new ArrayList<>();
        int cnt = 0;

        for (int i = 0; i < board.length(); i++) {
            if (board.charAt(i) == 'X') {
                cnt += 1;
            } else {
                if (cnt % 2 == 0) {
                    if (cnt >= 4) {
                        int num = cnt / 4;
                        for (int j = 0; j < num; j++) {
                            arr.add("AAAA");
                        }
                        cnt -= num * 4;
                    }
                    for (int j = 0; j < (cnt / 2); j++) {
                        arr.add("BB");
                        cnt -= 2;
                    }

                    if (i != board.length() - 1) {
                        arr.add(".");
                    }
                } else {
                    System.out.println(-1);
                    System.exit(0);
                }
            }
        }

        for (int i = 0; i < arr.size(); i++) {
            System.out.print(arr.get(i));
        }
    }
}
